src/main/java/com/example/photobucket/services/loginService/LoginService.java

```diff
@@ -0,0 +1,208 @@
+     * @throws UserDoesNotExistException if the username does not belong to an existing user
+     * @return true if the user is logged in, otherwise false
+     */
+    public boolean userLoggedIn(String username) {
+        if (!userExists(username)) throw new UserDoesNotExistException("User Does Not Exist");
+        User requestedUser = users.get(username);
+        return requestedUser.loggedIn();
+    }
+
+    /**
+     * Logs out an existing user of the Login Service
+     *
+     * @param username the username of a Login Service user
+     * @throws UserDoesNotExistException if the username does not belong to an existing user
+     */
+    public void logoutUser(String username) {
+        if (!userExists(username)) throw new UserDoesNotExistException("User Does Not Exist");
+        User requestedUser = users.get(username);
+        requestedUser.logOut();
+    }
+
+    /**
+     * Generates a "reset-password" token for a user of the Login Service
+     *
+     * @param username the username of a Login Service user
+     * @throws UserDoesNotExistException if the username does not belong to an existing user
+     */
+    public void generateToken(String username) {
+        if (!userExists(username)) throw new UserDoesNotExistException("User Does Not Exist");
+        User requestedUser = users.get(username);
+        requestedUser.generateToken();
+    }
+
+    /**
+     * Attempts to change the password for a user of the Login Service
+     *
+     * @param username the username of an existing user of the Login Service
+     * @param password the password of an existing user of the Login Service
+     * @param token the "reset-password" token to be used in the password change attempt
+     * @throws UserDoesNotExistException if the username does not belong to an existing user
+     * @throws IllegalChangePasswordTokenException if the "reset-password" token is invalid
+     *                                             (doesn't consist of 6 digits)
+     * @return true is the password change was successful, otherwise false
+     */
+    public boolean changePassword(String username, String token, String password) {
+        if (!userExists(username)) {
+            throw new UserDoesNotExistException("User Does Not Exist");
+        } else if (!token.matches("\\d{6}")) {
+            throw new IllegalChangePasswordTokenException("'reset-password' token needs to consist of 6 digits");
+        }
+        User requestedUser = users.get(username);
+        return requestedUser.changePassword(password, token);
+    }
+
+    /**
+     * A User of the Login Service
+     */
+    private static class User {
+
+        // The username of the user
+        public String username;
+
+        // The password of the user
+        public String password;
+
+        // The log-in state of the user
+        public boolean loggedIn;
+
+        // The "reset-password" token
+        public String token;
```

Comment:
Every user should not have a token since it is an unused field if the user does not request to change their password and stays that way. This should rather be handled by the login service.